<h1 align="center"> A Tale of Two cities</h1>
<h2 align="center">Clustering the Neighbourhoods of Mumbai and London</h2>

<p align = "center">Amanul Rahiman Shamshuddin Attar
<br>
<br>
30th January 2021
</p>

# Introduction

Mumbai and London are the most popular cities in the world. These two cities have major history in past. A lt has changed over the years and we now take a look at how cities have grown.

Mumbai and London are quite popular tourist and vacation destination for peoplearround the world. They are diverse and multicultural and offer a wide variety of experiences thet is widely sought after. We try to group the neighbourhoods of Mumbai and London respectively and  draw insights to what they look like now.

# Business Problem

The aim is to help tourists choose their destinations depending on the experiences that the neighbourhoods have to offer and what they would want to have. This also helps people make decisions if they are thinking about migrating to Mumbai or London or even if they want to relocate neighbourhoods within the city. Our findings will help stakeholders make informed decisions and address any concerns they have including the different kinds of cuisines, provision stores and what the city has to offer. 


# Data Description

We require geolocation data for both Mumbai and London. Postal codes in each city serve as a starting point. Using Postal codes we use can find out the neighbourhoods, boroughs, venues and their most popular venue categories.

## Mumbai

To derive our solution, We scrape our data from 
https://en.wikipedia.org/wiki/List_of_neighbourhoods_in_Mumbai

This wikipedia page has information about all the neighbourhoods, we limit it South Mumbai.

1. *borough* : Name of Neighbourhood
2. *town* : Name of borough
3. *latitude* : Latitude for Neighbourhood
4. *longitude* : Longitude for Neighbourhood

## London

To derive our solution, We scrape our data from https://en.wikipedia.org/wiki/List_of_areas_of_London

This wikipedia page has information about all the neighbourhoods, we limit it London.

1. *borough* : Name of Neighbourhood
2. *town* : Name of borough
3. *post_code* : Postal codes for London.

This wikipedia page lacks information about the geographical locations. To solve this problem we use ArcGIS API

### ArcGIS API

ArcGIS Online enables you to connect people, locations, and data using interactive maps. Work with smart, data-driven styles and intuitive analysis tools that deliver location intelligence. Share your insights with the world or specific groups. 

More specifically, we use ArcGIS to get the geo locations of the neighbourhoods of London. The following columns are added to our initial dataset which prepares our data. 

4. *latitude* : Latitude for Neighbourhood
5. *longitude* : Longitude for Neighbourhood

## Foursquare API Data

We will need data about different venues in different neighbourhoods of that specific borough. In order to gain that information we will use "Foursquare" locational information. Foursquare is a location data provider with information about all manner of venues and events within an area of interest. Such information includes venue names, locations, menus and even photos. As such, the foursquare location platform will be used as the sole data source since all the stated required information can be obtained through the API.

After finding the list of neighbourhoods, we then connect to the Foursquare API to gather information about venues inside each and every neighbourhood. For each neighbourhood, we have chosen the radius to be 1000 meters.

The data retrieved from Foursquare contained information of venues within a specified distance of the longitude and latitude of the postcodes. The information obtained per venue as follows:

1. *Neighbourhood* : Name of the Neighbourhood
2. *Neighbourhood Latitude* : Latitude of the Neighbourhood
3. *Neighbourhood Longitude* : Longitude of the Neighbourhood
4. *Venue* : Name of the Venue
5. *Venue Latitude* : Latitude of Venue
6. *Venue Longitude* : Longitude of Venue
7. *Venue Category* : Category of Venue


Based on all the information collected for both Mumbai and London, we have sufficient data to build our model. We cluster the neighbourhoods together based on similar venue categories. We then present our observations and findings. Using this data, our stakeholders can take the necessary decision.

# Methodology

We will be creating a model with the help of Python so we start off by importing all the required packages

In [7]:
import pandas as pd
import requests
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium 

#import k-means for the clustering stage
from sklearn.cluster import KMeans

The approach taken here is to explore each of the cities individually, plot the map to show the neighbourhoods being considered and then build our model by clustering all of the similar neighbourhoods together and finally plot the new map with the clustered neighbourhoods. We draw insights and then compare and discuss our findings.

# Exploring Mumbai

### Neighbourhood of London

We begin to start collecting and refining the data needed for the our business solution to work.

### Data Collection

To get the neighbourhoods in Mumbai, we start by scraping the list of areas of Mumbai wiki page.

In [8]:
url_mumbai = "https://en.wikipedia.org/wiki/List_of_neighbourhoods_in_Mumbai"
wiki_mumbai_url = requests.get(url_mumbai)
wiki_mumbai_url

<Response [200]>

Response 200 means that we are able to make the connection

In [9]:
wiki_mumbai_data = pd.read_html(wiki_mumbai_url.text)
wiki_mumbai_data

[                Area                 Location   Latitude  Longitude
 0             Amboli  Andheri,Western Suburbs  19.129300  72.843400
 1   Chakala, Andheri          Western Suburbs  19.111388  72.860833
 2         D.N. Nagar  Andheri,Western Suburbs  19.124085  72.831373
 3     Four Bungalows  Andheri,Western Suburbs  19.124714  72.827210
 4        Lokhandwala  Andheri,Western Suburbs  19.130815  72.829270
 ..               ...                      ...        ...        ...
 88             Parel             South Mumbai  18.990000  72.840000
 89      Gowalia Tank      Tardeo,South Mumbai  18.962450  72.809703
 90       Dava Bazaar             South Mumbai  18.946882  72.831362
 91           Dharavi                   Mumbai  19.040208  72.850850
 92             Thane                   Mumbai  19.200000  72.970000
 
 [93 rows x 4 columns]]

Scraping the webpage gives us all the tables present on the page. We need the 1st table, so selecting the 1st table.

In [10]:
wiki_mumbai_data = wiki_mumbai_data[0]
wiki_mumbai_data

Area                 Location   Latitude  Longitude
0             Amboli  Andheri,Western Suburbs  19.129300  72.843400
1   Chakala, Andheri          Western Suburbs  19.111388  72.860833
2         D.N. Nagar  Andheri,Western Suburbs  19.124085  72.831373
3     Four Bungalows  Andheri,Western Suburbs  19.124714  72.827210
4        Lokhandwala  Andheri,Western Suburbs  19.130815  72.829270
..               ...                      ...        ...        ...
88             Parel             South Mumbai  18.990000  72.840000
89      Gowalia Tank      Tardeo,South Mumbai  18.962450  72.809703
90       Dava Bazaar             South Mumbai  18.946882  72.831362
91           Dharavi                   Mumbai  19.040208  72.850850
92             Thane                   Mumbai  19.200000  72.970000

[93 rows x 4 columns]

### Data Preprocessing

we remove the spaces in the column titels and then we add _ between words

In [11]:
wiki_mumbai_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
wiki_mumbai_data

Area                 Location   Latitude  Longitude
0             Amboli  Andheri,Western Suburbs  19.129300  72.843400
1   Chakala, Andheri          Western Suburbs  19.111388  72.860833
2         D.N. Nagar  Andheri,Western Suburbs  19.124085  72.831373
3     Four Bungalows  Andheri,Western Suburbs  19.124714  72.827210
4        Lokhandwala  Andheri,Western Suburbs  19.130815  72.829270
..               ...                      ...        ...        ...
88             Parel             South Mumbai  18.990000  72.840000
89      Gowalia Tank      Tardeo,South Mumbai  18.962450  72.809703
90       Dava Bazaar             South Mumbai  18.946882  72.831362
91           Dharavi                   Mumbai  19.040208  72.850850
92             Thane                   Mumbai  19.200000  72.970000

[93 rows x 4 columns]

We see that few columns have no '_' between the words despite applying our function meaning that there are special characters

### Feature Selection

In [12]:
df1 = wiki_mumbai_data
df1.head()

Area                 Location   Latitude  Longitude
0            Amboli  Andheri,Western Suburbs  19.129300  72.843400
1  Chakala, Andheri          Western Suburbs  19.111388  72.860833
2        D.N. Nagar  Andheri,Western Suburbs  19.124085  72.831373
3    Four Bungalows  Andheri,Western Suburbs  19.124714  72.827210
4       Lokhandwala  Andheri,Western Suburbs  19.130815  72.829270

let's rename the area, location column and the london borough to something simpler

In [36]:
df1.columns = ["borough", "town","latitude","longitude"]
df1

borough                      town   latitude  longitude
52         Agripada              South Mumbai  18.977700  72.827300
53   Altamount Road              South Mumbai  18.968100  72.809500
54       Bhuleshwar              South Mumbai  18.950000  72.830000
55     Breach Candy              South Mumbai  18.967000  72.805000
56  Carmichael Road              South Mumbai  18.972200  72.811300
57            Cavel              South Mumbai  18.947400  72.827200
58       Churchgate              South Mumbai  18.930000  72.820000
59     Cotton Green              South Mumbai  18.986209  72.844076
60     Cuffe Parade              South Mumbai  18.910000  72.810000
61     Cumbala Hill              South Mumbai  18.965833  72.805833
62      Currey Road              South Mumbai  18.994000  72.833000
63       Dhobitalao              South Mumbai  18.943300  72.828600
64           Dongri              South Mumbai  19.283333  72.783333
65       Kala Ghoda              South Mumbai  18.930700  72.833100
66     Kemps Corner              South Mumbai  18.962900  72.805400
67      Lower Parel              South Mumbai  18.995278  72.830000
68        Mahalaxmi              South Mumbai  18.983333  72.800000
69            Mahim              South Mumbai  19.035000  72.840000
70     Malabar Hill              South Mumbai  18.950000  72.795000
71     Marine Drive              South Mumbai  18.944000  72.823000
72     Marine Lines              South Mumbai  18.944700  72.824400
73   Mumbai Central              South Mumbai  18.969700  72.819400
74    Nariman Point              South Mumbai  18.926000  72.823000
75       Prabhadevi              South Mumbai  19.016600  72.829500
76             Sion              South Mumbai  19.040000  72.860000
77       Walkeshwar              South Mumbai  18.947596  72.795957
78            Worli              South Mumbai  19.000000  72.815000
79    C.G.S. colony   Antop Hill,South Mumbai  19.016378  72.856629
80      Dagdi Chawl      Byculla,South Mumbai  18.977129  72.829131
81       Navy Nagar       Colaba,South Mumbai  18.901200  72.810100
82     Hindu colony        Dadar,South Mumbai  19.020841  19.020841
83   Ballard Estate         Fort,South Mumbai  18.950000  72.840000
84     Chira Bazaar    Kalbadevi,South Mumbai  18.948140  72.825462
85       Fanas Wadi    Kalbadevi,South Mumbai  18.951811  72.825309
86      Chor Bazaar  Kamathipura,South Mumbai  18.960321  72.827176
87          Matunga              South Mumbai  19.017980  72.844763
88            Parel              South Mumbai  18.990000  72.840000
89     Gowalia Tank       Tardeo,South Mumbai  18.962450  72.809703
90      Dava Bazaar              South Mumbai  18.946882  72.831362

Let's remove the Square brackets [ ] and numbers from the borough column

In [37]:
df1['borough'] = df1['borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1

borough                      town   latitude  longitude
52         Agripada              South Mumbai  18.977700  72.827300
53   Altamount Road              South Mumbai  18.968100  72.809500
54       Bhuleshwar              South Mumbai  18.950000  72.830000
55     Breach Candy              South Mumbai  18.967000  72.805000
56  Carmichael Road              South Mumbai  18.972200  72.811300
57            Cavel              South Mumbai  18.947400  72.827200
58       Churchgate              South Mumbai  18.930000  72.820000
59     Cotton Green              South Mumbai  18.986209  72.844076
60     Cuffe Parade              South Mumbai  18.910000  72.810000
61     Cumbala Hill              South Mumbai  18.965833  72.805833
62      Currey Road              South Mumbai  18.994000  72.833000
63       Dhobitalao              South Mumbai  18.943300  72.828600
64           Dongri              South Mumbai  19.283333  72.783333
65       Kala Ghoda              South Mumbai  18.930700  72.833100
66     Kemps Corner              South Mumbai  18.962900  72.805400
67      Lower Parel              South Mumbai  18.995278  72.830000
68        Mahalaxmi              South Mumbai  18.983333  72.800000
69            Mahim              South Mumbai  19.035000  72.840000
70     Malabar Hill              South Mumbai  18.950000  72.795000
71     Marine Drive              South Mumbai  18.944000  72.823000
72     Marine Lines              South Mumbai  18.944700  72.824400
73   Mumbai Central              South Mumbai  18.969700  72.819400
74    Nariman Point              South Mumbai  18.926000  72.823000
75       Prabhadevi              South Mumbai  19.016600  72.829500
76             Sion              South Mumbai  19.040000  72.860000
77       Walkeshwar              South Mumbai  18.947596  72.795957
78            Worli              South Mumbai  19.000000  72.815000
79    C.G.S. colony   Antop Hill,South Mumbai  19.016378  72.856629
80      Dagdi Chawl      Byculla,South Mumbai  18.977129  72.829131
81       Navy Nagar       Colaba,South Mumbai  18.901200  72.810100
82     Hindu colony        Dadar,South Mumbai  19.020841  19.020841
83   Ballard Estate         Fort,South Mumbai  18.950000  72.840000
84     Chira Bazaar    Kalbadevi,South Mumbai  18.948140  72.825462
85       Fanas Wadi    Kalbadevi,South Mumbai  18.951811  72.825309
86      Chor Bazaar  Kamathipura,South Mumbai  18.960321  72.827176
87          Matunga              South Mumbai  19.017980  72.844763
88            Parel              South Mumbai  18.990000  72.840000
89     Gowalia Tank       Tardeo,South Mumbai  18.962450  72.809703
90      Dava Bazaar              South Mumbai  18.946882  72.831362

take dimension of the data frame

In [38]:
df1.shape

(39, 4)

We currently have 533 records and 3 columns of our data. It's time to perform Feature Engineering

### Feature Engineering

We can only focusing on the neighbourhoods of South Mumbai, so performing the changes

In [39]:
df1 = df1[df1['town'].str.contains('South Mumbai')]
df1

borough                      town   latitude  longitude
52         Agripada              South Mumbai  18.977700  72.827300
53   Altamount Road              South Mumbai  18.968100  72.809500
54       Bhuleshwar              South Mumbai  18.950000  72.830000
55     Breach Candy              South Mumbai  18.967000  72.805000
56  Carmichael Road              South Mumbai  18.972200  72.811300
57            Cavel              South Mumbai  18.947400  72.827200
58       Churchgate              South Mumbai  18.930000  72.820000
59     Cotton Green              South Mumbai  18.986209  72.844076
60     Cuffe Parade              South Mumbai  18.910000  72.810000
61     Cumbala Hill              South Mumbai  18.965833  72.805833
62      Currey Road              South Mumbai  18.994000  72.833000
63       Dhobitalao              South Mumbai  18.943300  72.828600
64           Dongri              South Mumbai  19.283333  72.783333
65       Kala Ghoda              South Mumbai  18.930700  72.833100
66     Kemps Corner              South Mumbai  18.962900  72.805400
67      Lower Parel              South Mumbai  18.995278  72.830000
68        Mahalaxmi              South Mumbai  18.983333  72.800000
69            Mahim              South Mumbai  19.035000  72.840000
70     Malabar Hill              South Mumbai  18.950000  72.795000
71     Marine Drive              South Mumbai  18.944000  72.823000
72     Marine Lines              South Mumbai  18.944700  72.824400
73   Mumbai Central              South Mumbai  18.969700  72.819400
74    Nariman Point              South Mumbai  18.926000  72.823000
75       Prabhadevi              South Mumbai  19.016600  72.829500
76             Sion              South Mumbai  19.040000  72.860000
77       Walkeshwar              South Mumbai  18.947596  72.795957
78            Worli              South Mumbai  19.000000  72.815000
79    C.G.S. colony   Antop Hill,South Mumbai  19.016378  72.856629
80      Dagdi Chawl      Byculla,South Mumbai  18.977129  72.829131
81       Navy Nagar       Colaba,South Mumbai  18.901200  72.810100
82     Hindu colony        Dadar,South Mumbai  19.020841  19.020841
83   Ballard Estate         Fort,South Mumbai  18.950000  72.840000
84     Chira Bazaar    Kalbadevi,South Mumbai  18.948140  72.825462
85       Fanas Wadi    Kalbadevi,South Mumbai  18.951811  72.825309
86      Chor Bazaar  Kamathipura,South Mumbai  18.960321  72.827176
87          Matunga              South Mumbai  19.017980  72.844763
88            Parel              South Mumbai  18.990000  72.840000
89     Gowalia Tank       Tardeo,South Mumbai  18.962450  72.809703
90      Dava Bazaar              South Mumbai  18.946882  72.831362

In [40]:
df1.shape

(39, 4)

## Geolocations of the London Neighbourhoods

### ArcGis API

We need to get the geographical co-ordinates for the neighbourhoods to plot out map. We will use the arcgis package to do so. 

Arcgis doesn't have a limitation on the number of API calls made so it fits our use case perfectly.

In [41]:
pip install arcgis

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\users\amana\appdata\local\programs\python\python38\python.exe -m pip install --upgrade pip' command.


Getting th geocodes for Mumbai to help visualize it on the map

In [42]:
from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()

### Co-ordinates for Mumbai

Getting the geocode for Mumbai to help visualize it on the map

In [43]:
mumbai = geocode(address='Mumbai,India')[0]
mumbai_lng_coords = mumbai['location']['x']
mumbai_lat_coords = mumbai['location']['y']
mumbai_lng_coords

72.83483000000007

In [44]:
mumbai_lat_coords

18.940170000000023

# Visualize the Map of Mumbai

To help visualize the Map of Mumbai and the neighbourhoods in Mumbai, we make use of the folium package.

In [45]:
# Creating the map of Mumbai
map_Mumbai = folium.Map(location=[mumbai_lat_coords, mumbai_lng_coords], zoom_start=12)
map_Mumbai

# adding markers to map
for latitude, longitude, borough, town in zip(df1['latitude'], df1['longitude'], df1['borough'], df1['town']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Mumbai)  
    
map_Mumbai

### Venues in Mumbai

To proceed with the next part, we need to define Foursquare API credentials.

Using Foursquare API, we are able to get the venue and venue categories around each neighbourhood in Mumbai.

In [55]:
CLIENT_ID = 'XOCJVA41PPQHX5ZVPCWPWESBQEJT5J05QUJVEQXPPFROB3YJ'
CLIENT_SECRET = 'KHQOY5O3TIKTEOGRFGQBJ5JUMBLBCLZJMIWGJPCA5NTXABIF'
VERSION = '20210101'

Defining a function to get the neraby venues in the neighbourhood. This will help us get venue categories which is important for our analysis

In [56]:
LIMIT=400

def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            LIMIT
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Getting the venues in London

In [57]:
venues_in_Mumbai = getNearbyVenues(df1['borough'], df1['latitude'], df1['longitude'])

Agripada
Altamount Road
Bhuleshwar
Breach Candy
Carmichael Road
Cavel
Churchgate
Cotton Green
Cuffe Parade
Cumbala Hill
Currey Road
Dhobitalao
Dongri
Kala Ghoda
Kemps Corner
Lower Parel
Mahalaxmi
Mahim
Malabar Hill
Marine Drive
Marine Lines
Mumbai Central
Nariman Point
Prabhadevi
Sion
Walkeshwar
Worli
C.G.S. colony
Dagdi Chawl
Navy Nagar
Hindu colony
Ballard Estate
Chira Bazaar
Fanas Wadi
Chor Bazaar
Matunga
Parel
Gowalia Tank
Dava Bazaar


Sampling our data

In [58]:
venues_in_Mumbai.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0      Agripada                 18.9777                  72.8273   
1      Agripada                 18.9777                  72.8273   
2      Agripada                 18.9777                  72.8273   
3      Agripada                 18.9777                  72.8273   
4      Agripada                 18.9777                  72.8273   

                                               Venue     Venue Category  
0                                            Celejor             Bakery  
1                                   Tote On The Turf          Nightclub  
2  Mahalaxmi Race Course (Royal Western India Tur...         Club House  
3                               Bhau Daji Lad Museum     History Museum  
4                                     Persian Darbar  Indian Restaurant

In [59]:
venues_in_Mumbai.shape

(1707, 5)

Wow, we have scraped together 1707 records for venues. This will definitely make the clustering interesting.


### Grouping by Venue Categories
We need to now see how many Venue Categories are there for further processing

In [60]:
venues_in_Mumbai.groupby("Venue Category").max()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                               
American Restaurant     Dhobitalao               18.950000   
Antique Shop           Chor Bazaar               18.960321   
Aquarium              Marine Lines               18.951811   
Arcade                       Worli               19.017980   
Art Gallery                  Worli               19.000000   
...                            ...                     ...   
Wine Bar                     Worli               19.000000   
Wine Shop               Dhobitalao               18.950000   
Women's Store        Nariman Point               19.017980   
Yoga Studio           Gowalia Tank               19.016378   
Zoo                    Dagdi Chawl               18.977700   

                     Neighbourhood Longitude                          Venue  
Venue Category                                                               
American Restaurant                72.840000                         Subway  
Antique Shop                       72.827176  Chor Bazaar (Thieves' Market)  
Aquarium                           72.827200          Taraporewala Aquarium  
Arcade                             72.844763     Worli Sea Face Kebab rolls  
Art Gallery                        72.833100                     SaffronArt  
...                                      ...                            ...  
Wine Bar                           72.823000               Vinoteca by Sula  
Wine Shop                          72.840000                   Peekay Wines  
Women's Store                      72.844763                       Westside  
Yoga Studio                        72.856629     Moksh - The Wellness Place  
Zoo                                72.829131      Jijamata Udyan - Ranibaug  

[156 rows x 4 columns]

We can see 156 records, just goes to show how diverse and interesting the place is.

### One Hot Encoding 
We need to Encode our venue categories to get a better result for our clustering

In [62]:
Mumbai_venue_cat = pd.get_dummies(venues_in_Mumbai[['Venue Category']], prefix="", prefix_sep="")
Mumbai_venue_cat

American Restaurant  Antique Shop  Aquarium  Arcade  Art Gallery  \
0                       0             0         0       0            0   
1                       0             0         0       0            0   
2                       0             0         0       0            0   
3                       0             0         0       0            0   
4                       0             0         0       0            0   
...                   ...           ...       ...     ...          ...   
1702                    0             0         0       0            0   
1703                    0             0         0       0            0   
1704                    0             0         0       0            0   
1705                    0             0         0       0            0   
1706                    0             0         0       0            0   

      Arts & Crafts Store  Asian Restaurant  Athletics & Sports  BBQ Joint  \
0                       0                 0                   0          0   
1                       0                 0                   0          0   
2                       0                 0                   0          0   
3                       0                 0                   0          0   
4                       0                 0                   0          0   
...                   ...               ...                 ...        ...   
1702                    0                 0                   0          0   
1703                    0                 0                   0          0   
1704                    0                 0                   0          0   
1705                    0                 0                   0          0   
1706                    0                 0                   0          0   

      Bakery  ...  Trail  Train  Train Station  Vegetarian / Vegan Restaurant  \
0          1  ...      0      0              0                              0   
1          0  ...      0      0              0                              0   
2          0  ...      0      0              0                              0   
3          0  ...      0      0              0                              0   
4          0  ...      0      0              0                              0   
...      ...  ...    ...    ...            ...                            ...   
1702       0  ...      0      0              0                              0   
1703       0  ...      0      0              0                              0   
1704       0  ...      0      0              0                              0   
1705       0  ...      0      0              0                              0   
1706       0  ...      0      0              0                              0   

      Video Store  Wine Bar  Wine Shop  Women's Store  Yoga Studio  Zoo  
0               0         0          0              0            0    0  
1               0         0          0              0            0    0  
2               0         0          0              0            0    0  
3               0         0          0              0            0    0  
4               0         0          0              0            0    0  
...           ...       ...        ...            ...          ...  ...  
1702            0         0          0              0            0    0  
1703            0         0          0              0            0    0  
1704            0         0          0              0            0    0  
1705            0         0          0              0            0    0  
1706            0         0          0              0            0    0  

[1707 rows x 156 columns]

Adding Neighbourhood into the mix.

In [64]:
Mumbai_venue_cat['Neighbourhood'] = venues_in_Mumbai['Neighbourhood']

# moving neghbourhood colun to the first column
fixed_columns = [Mumbai_venue_cat.columns[-1]] + list(Mumbai_venue_cat.columns[:-1])
Mumbai_venue_cat = Mumbai_venue_cat[fixed_columns]

Mumbai_venue_cat.head()

Neighbourhood  American Restaurant  Antique Shop  Aquarium  Arcade  \
0      Agripada                    0             0         0       0   
1      Agripada                    0             0         0       0   
2      Agripada                    0             0         0       0   
3      Agripada                    0             0         0       0   
4      Agripada                    0             0         0       0   

   Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0            0                    0                 0                   0   
1            0                    0                 0                   0   
2            0                    0                 0                   0   
3            0                    0                 0                   0   
4            0                    0                 0                   0   

   BBQ Joint  ...  Trail  Train  Train Station  Vegetarian / Vegan Restaurant  \
0          0  ...      0      0              0                              0   
1          0  ...      0      0              0                              0   
2          0  ...      0      0              0                              0   
3          0  ...      0      0              0                              0   
4          0  ...      0      0              0                              0   

   Video Store  Wine Bar  Wine Shop  Women's Store  Yoga Studio  Zoo  
0            0         0          0              0            0    0  
1            0         0          0              0            0    0  
2            0         0          0              0            0    0  
3            0         0          0              0            0    0  
4            0         0          0              0            0    0  

[5 rows x 157 columns]

### Venue categories mean value
We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [65]:
Mumbai_grouped = Mumbai_venue_cat.groupby('Neighbourhood').mean().reset_index()
Mumbai_grouped.head()

Neighbourhood  American Restaurant  Antique Shop  Aquarium    Arcade  \
0        Agripada             0.000000           0.0       0.0  0.000000   
1  Altamount Road             0.000000           0.0       0.0  0.000000   
2  Ballard Estate             0.034483           0.0       0.0  0.000000   
3      Bhuleshwar             0.019608           0.0       0.0  0.019608   
4    Breach Candy             0.000000           0.0       0.0  0.000000   

   Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0          0.0                  0.0          0.037037                 0.0   
1          0.0                  0.0          0.000000                 0.0   
2          0.0                  0.0          0.000000                 0.0   
3          0.0                  0.0          0.000000                 0.0   
4          0.0                  0.0          0.000000                 0.0   

   BBQ Joint  ...  Trail  Train  Train Station  Vegetarian / Vegan Restaurant  \
0   0.000000  ...    0.0    0.0       0.000000                       0.000000   
1   0.000000  ...    0.0    0.0       0.000000                       0.024390   
2   0.034483  ...    0.0    0.0       0.000000                       0.000000   
3   0.019608  ...    0.0    0.0       0.019608                       0.000000   
4   0.000000  ...    0.0    0.0       0.000000                       0.018519   

   Video Store  Wine Bar  Wine Shop  Women's Store  Yoga Studio       Zoo  
0          0.0       0.0   0.000000       0.000000     0.000000  0.037037  
1          0.0       0.0   0.000000       0.012195     0.012195  0.000000  
2          0.0       0.0   0.034483       0.000000     0.000000  0.000000  
3          0.0       0.0   0.019608       0.000000     0.000000  0.000000  
4          0.0       0.0   0.000000       0.018519     0.000000  0.000000  

[5 rows x 157 columns]

Let's make a function to get the top most common venue categories

In [66]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods.

Creating a function to label the columns of the venue correctly

In [67]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


### Top venue categories

Getting the top venue categories in London

In [69]:
# create a new dataframe for Mumbai
neighborhoods_venues_sorted_mumbai = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_mumbai['Neighbourhood'] = Mumbai_grouped['Neighbourhood']

for ind in np.arange(Mumbai_grouped.shape[0]):
    neighborhoods_venues_sorted_mumbai.iloc[ind, 1:] = return_most_common_venues(Mumbai_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_mumbai.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0        Agripada     Indian Restaurant                Bakery   
1  Altamount Road    Chinese Restaurant                  Café   
2  Ballard Estate     Indian Restaurant       Harbor / Marina   
3      Bhuleshwar     Indian Restaurant                Bakery   
4    Breach Candy                  Café                Bakery   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Gym            Restaurant           Coffee Shop   
1                   Bar     Indian Restaurant                Bakery   
2                  Café            Smoke Shop        Ice Cream Shop   
3                  Café    Chinese Restaurant  Fast Food Restaurant   
4           Coffee Shop     Indian Restaurant        Sandwich Place   

    6th Most Common Venue 7th Most Common Venue      8th Most Common Venue  \
0                     Zoo        History Museum                Pizza Place   
1    Fast Food Restaurant           Pizza Place             Sandwich Place   
2  Furniture / Home Store             Rest Area  Middle Eastern Restaurant   
3            Dessert Shop         Jewelry Store                Music Store   
4            Dessert Shop           Pizza Place         Chinese Restaurant   

  9th Most Common Venue     10th Most Common Venue  
0          Cupcake Shop       Fast Food Restaurant  
1           Coffee Shop         Salon / Barbershop  
2                Market                      Hotel  
3             Multiplex  Middle Eastern Restaurant  
4                  Park                        Bar

## Model Building

### K Mean
Let's cluster the city of Mumbai to roughly 5 to make it easier to analyze

We use K Means clustering technique to do so.


In [78]:
# set number of clusters
k_num_clusters= 5

Mumbai_grouped_clustering = Mumbai_grouped.drop('Neighbourhood',1)

# run k-means clustering
kmeans_mumbai = KMeans(n_clusters=k_num_cluster, random_state=0).fit(Mumbai_grouped_clustering)
kmeans_mumbai

KMeans(n_clusters=5, random_state=0)

### Labelling Clustered Data

In [79]:
kmeans_mumbai.labels_

array([3, 0, 3, 3, 0, 0, 0, 3, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 2, 3, 0, 3,
       0, 0, 1, 3, 3, 3, 3, 3, 0, 0, 4, 0, 3, 3, 3, 0])

So our model has labeled the city

In [80]:
neighborhoods_venues_sorted_mumbai.insert(0,'Cluster Labels',kmeans_mumbai.labels_ +1)

ValueError: cannot insert Cluster Labels, already exists

Join df1 with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [81]:
mumbai_data = df1

mumbai_data = mumbai_data.join(neighborhoods_venues_sorted_mumbai.set_index('Neighbourhood'), on = 'borough')

mumbai_data.head()

borough          town  latitude  longitude  Cluster Labels  \
52         Agripada  South Mumbai   18.9777    72.8273             4.0   
53   Altamount Road  South Mumbai   18.9681    72.8095             1.0   
54       Bhuleshwar  South Mumbai   18.9500    72.8300             4.0   
55     Breach Candy  South Mumbai   18.9670    72.8050             1.0   
56  Carmichael Road  South Mumbai   18.9722    72.8113             1.0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
52     Indian Restaurant                Bakery                   Gym   
53    Chinese Restaurant                  Café                   Bar   
54     Indian Restaurant                Bakery                  Café   
55                  Café                Bakery           Coffee Shop   
56  Fast Food Restaurant    Chinese Restaurant     Indian Restaurant   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
52            Restaurant           Coffee Shop                   Zoo   
53     Indian Restaurant                Bakery  Fast Food Restaurant   
54    Chinese Restaurant  Fast Food Restaurant          Dessert Shop   
55     Indian Restaurant        Sandwich Place          Dessert Shop   
56                   Bar                Bakery        Ice Cream Shop   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
52        History Museum           Pizza Place          Cupcake Shop   
53           Pizza Place        Sandwich Place           Coffee Shop   
54         Jewelry Store           Music Store             Multiplex   
55           Pizza Place    Chinese Restaurant                  Park   
56        Sandwich Place           Pizza Place         Shopping Mall   

       10th Most Common Venue  
52       Fast Food Restaurant  
53         Salon / Barbershop  
54  Middle Eastern Restaurant  
55                        Bar  
56                       Park

Drop all the NaN values to prevent data skew

In [82]:
mumbai_data_nonan = mumbai_data.dropna(subset=['Cluster Labels'])

### Visualizing the clustered neighbourhood
Let's plot the clusters

In [84]:
map_clusters_mumbai = folium.Map(location=[mumbai_lat_coords,mumbai_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(mumbai_data_nonan['latitude'], mumbai_data_nonan['longitude'], mumbai_data_nonan['borough'], mumbai_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters_mumbai)
        
map_clusters_mumbai

### Examining our Clusters

Cluster 1

In [85]:
mumbai_data_nonan.loc[mumbai_data_nonan['Cluster Labels'] == 1, mumbai_data_nonan.columns[[1] + list(range(5, mumbai_data_nonan.shape[1]))]]

town 1st Most Common Venue 2nd Most Common Venue  \
53             South Mumbai    Chinese Restaurant                  Café   
55             South Mumbai                  Café                Bakery   
56             South Mumbai  Fast Food Restaurant    Chinese Restaurant   
58             South Mumbai  Fast Food Restaurant    Italian Restaurant   
59             South Mumbai         Train Station        Ice Cream Shop   
60             South Mumbai                  Park     Indian Restaurant   
61             South Mumbai                  Café                Bakery   
62             South Mumbai     Indian Restaurant                  Café   
66             South Mumbai     Indian Restaurant                  Café   
67             South Mumbai     Indian Restaurant                  Café   
73             South Mumbai  Fast Food Restaurant     Indian Restaurant   
74             South Mumbai     Indian Restaurant                  Café   
78             South Mumbai        Ice Cream Shop          Dessert Shop   
79  Antop Hill,South Mumbai  Fast Food Restaurant           Coffee Shop   
88             South Mumbai                 Plaza            Restaurant   
89      Tardeo,South Mumbai     Indian Restaurant  Fast Food Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
53                   Bar     Indian Restaurant                Bakery   
55           Coffee Shop     Indian Restaurant        Sandwich Place   
56     Indian Restaurant                   Bar                Bakery   
58                 Hotel     Indian Restaurant        Ice Cream Shop   
59  Fast Food Restaurant             Multiplex                 Plaza   
60      Department Store    Italian Restaurant      Asian Restaurant   
61           Coffee Shop     Indian Restaurant        Sandwich Place   
62           Coffee Shop         Shopping Mall           Pizza Place   
66                Bakery                  Park            Restaurant   
67                   Pub            Restaurant        Clothing Store   
73        Ice Cream Shop                 Hotel      Department Store   
74  Fast Food Restaurant    Italian Restaurant           Coffee Shop   
78    Chinese Restaurant      Asian Restaurant           Bus Station   
79                  Café     Convention Center                 Trail   
88                 Hotel  Fast Food Restaurant     Indian Restaurant   
89                  Café           Coffee Shop                Bakery   

   6th Most Common Venue 7th Most Common Venue          8th Most Common Venue  \
53  Fast Food Restaurant           Pizza Place                 Sandwich Place   
55          Dessert Shop           Pizza Place             Chinese Restaurant   
56        Ice Cream Shop        Sandwich Place                    Pizza Place   
58           Coffee Shop            Restaurant                        Theater   
59           Snack Place                 Hotel  Vegetarian / Vegan Restaurant   
60         Grocery Store      Basketball Court                          Beach   
61           Pizza Place    Salon / Barbershop           Fast Food Restaurant   
62  Fast Food Restaurant            Restaurant                 Clothing Store   
66  Fast Food Restaurant    Salon / Barbershop                 Sandwich Place   
67           Coffee Shop    Italian Restaurant                          Hotel   
73    Chinese Restaurant                Bakery                    Coffee Shop   
74            Restaurant        Ice Cream Shop                            Bar   
78     Indian Restaurant                   Spa                          Diner   
79         Movie Theater          Cupcake Shop                         Garden   
88         Train Station    Chinese Restaurant                      Roof Deck   
89           Snack Place     Electronics Store                 Sandwich Place   

   9th Most Common Venue           10th Most Common Venue  
53           Coffee Shop               Salon / Barbershop  
55                  Park

Cluster 2

In [86]:
mumbai_data_nonan.loc[mumbai_data_nonan['Cluster Labels'] == 2, mumbai_data_nonan.columns[[1] + list(range(5, mumbai_data_nonan.shape[1]))]]

town 1st Most Common Venue 2nd Most Common Venue  \
68  South Mumbai                Arcade                   Zoo   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
68        Farmers Market     Food & Drink Shop                  Food   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
68           Flower Shop           Flea Market           Fish Market   

   9th Most Common Venue 10th Most Common Venue  
68                 Field   Fast Food Restaurant

Cluster 3

In [87]:
mumbai_data_nonan.loc[mumbai_data_nonan['Cluster Labels'] == 3, mumbai_data_nonan.columns[[1] + list(range(5, mumbai_data_nonan.shape[1]))]]

town 1st Most Common Venue 2nd Most Common Venue  \
64  South Mumbai                 Beach            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
64                Resort           Bus Station     Indian Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
64          Dance Studio        Farmers Market           Flower Shop   

   9th Most Common Venue 10th Most Common Venue  
64           Flea Market            Fish Market

Cluster 4

In [88]:
mumbai_data_nonan.loc[mumbai_data_nonan['Cluster Labels'] == 4, mumbai_data_nonan.columns[[1] + list(range(5, mumbai_data_nonan.shape[1]))]]

town 1st Most Common Venue  \
52              South Mumbai     Indian Restaurant   
54              South Mumbai     Indian Restaurant   
57              South Mumbai     Indian Restaurant   
63              South Mumbai     Indian Restaurant   
65              South Mumbai     Indian Restaurant   
69              South Mumbai     Indian Restaurant   
70              South Mumbai     Indian Restaurant   
71              South Mumbai     Indian Restaurant   
72              South Mumbai     Indian Restaurant   
75              South Mumbai     Indian Restaurant   
76              South Mumbai     Indian Restaurant   
77              South Mumbai     Indian Restaurant   
80      Byculla,South Mumbai     Indian Restaurant   
83         Fort,South Mumbai     Indian Restaurant   
84    Kalbadevi,South Mumbai     Indian Restaurant   
85    Kalbadevi,South Mumbai     Indian Restaurant   
86  Kamathipura,South Mumbai     Indian Restaurant   
87              South Mumbai     Indian Restaurant   
90              South Mumbai     Indian Restaurant   

            2nd Most Common Venue          3rd Most Common Venue  \
52                         Bakery                            Gym   
54                         Bakery                           Café   
57                         Bakery                           Food   
63                  Train Station                           Café   
65                           Café                            Bar   
69             Chinese Restaurant                           Café   
70                 Ice Cream Shop                           Park   
71                           Café             Chinese Restaurant   
72                           Café                  Jewelry Store   
75                    Coffee Shop                           Café   
76  Vegetarian / Vegan Restaurant             Seafood Restaurant   
77                 Ice Cream Shop  Vegetarian / Vegan Restaurant   
80                    Coffee Shop                         Bakery   
83                Harbor / Marina                           Café   
84                         Bakery                  Jewelry Store   
85                           Café           Fast Food Restaurant   
86                   Dessert Shop           Fast Food Restaurant   
87                    Coffee Shop             Chinese Restaurant   
90                           Café           Fast Food Restaurant   

   4th Most Common Venue 5th Most Common Venue   6th Most Common Venue  \
52            Restaurant           Coffee Shop                     Zoo   
54    Chinese Restaurant  Fast Food Restaurant            Dessert Shop   
57    Chinese Restaurant            Restaurant    Fast Food Restaurant   
63                Bakery                   Bar           Jewelry Store   
65    Seafood Restaurant           Coffee Shop      Chinese Restaurant   
69           Coffee Shop                 Beach    Fast Food Restaurant   
70                   Gym          Dessert Shop             Coffee Shop   
71           Bridal Shop        Cricket Ground           Train Station   
72                   Bar    Chinese Restaurant             Bridal Shop   
75  Fast Food Restaurant    Chinese Restaurant                  Bakery   
76         Train Station             Juice Bar                   Plaza   
77            Lighthouse          Dessert Shop                     Gym   
80            Restaurant                   Gym                     Zoo   
83            Smoke Shop        Ice Cream Shop  Furniture / Home Store   
84              Boutique            Restaurant           Train Station   
85             Juice Bar    Chinese Restaurant      Italian Restaurant   
86             BBQ Joint    Chinese Restaurant        Department Store   
87  Fast Food Restaurant         Movie Theater          Breakfast Spot   
90                Bakery                   Bar      Chinese Restaurant   

            7th Most Common Venue      8th Most Common Venue  \
52                 History Museum    

Cluster 5

In [90]:
mumbai_data_nonan.loc[mumbai_data_nonan['Cluster Labels'] == 5, mumbai_data_nonan.columns[[1] + list(range(5, mumbai_data_nonan.shape[1]))]]

town 1st Most Common Venue 2nd Most Common Venue  \
81  Colaba,South Mumbai           Golf Course      Asian Restaurant   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
81  General Entertainment                 Beach        Farmers Market   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
81                  Food           Flower Shop           Flea Market   

   9th Most Common Venue 10th Most Common Venue  
81           Fish Market                  Field



---



# Exploring London

### Neighbourhoods of London

We begin to start collecting and refining the data needed for the our business solution to work.

### Data Collection

To get the neighbourhoods in london, we start by scraping the list of areas of london wiki page.

In [91]:
url_london = "https://en.wikipedia.org/wiki/List_of_areas_of_London"
wiki_london_url = requests.get(url_london)
wiki_london_url

<Response [200]>

Response 200 means that we are able to make the connection

In [92]:
wiki_london_data = pd.read_html(wiki_london_url.text)
wiki_london_data

[                                                   0
 0  Map all coordinates in "Category:Areas of Lond...
 1                 Download coordinates as: KML · GPX,
             Location                     London borough       Post town  \
 0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
 1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
 2          Addington                         Croydon[8]         CROYDON   
 3         Addiscombe                         Croydon[8]         CROYDON   
 4        Albany Park                             Bexley  BEXLEY, SIDCUP   
 ..               ...                                ...             ...   
 526         Woolwich                          Greenwich          LONDON   
 527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
 528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
 529          Yeading                         Hillingdon           HAYES   
 

Scraping the webpage gives us all the tables present on the page. We need the 2nd table, so selecting the 2nd table.

In [93]:
wiki_london_data = wiki_london_data[1]
wiki_london_data

Location                     London borough       Post town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
526         Woolwich                          Greenwich          LONDON   
527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
529          Yeading                         Hillingdon           HAYES   
530         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS grid ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
526              SE18       020    TQ435795  
527               KT4       020    TQ225655  
528               W12       020    TQ225815  
529               UB4       020    TQ115825  
530               UB7       020    TQ063804  

[531 rows x 6 columns]

### Data Preprocessing

we remove the spaces in the column titles and then we add _ between words.

In [94]:
wiki_london_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
wiki_london_data

Location                     London borough       Post_town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
526         Woolwich                          Greenwich          LONDON   
527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
529          Yeading                         Hillingdon           HAYES   
530         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS_grid_ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
526              SE18       020    TQ435795  
527               KT4       020    TQ225655  
528               W12       020    TQ225815  
529               UB4       020    TQ115825  
530               UB7       020    TQ063804  

[531 rows x 6 columns]

We see that few columns have no '_' between the words despite applying our function meaning that there are special characters

### Feature Selection

We need only the boroughs, Postal codes, Post town for further steps. We can drop the locations, dial codes and OS grid.

In [95]:
df1 = wiki_london_data.drop( [ wiki_london_data.columns[0], wiki_london_data.columns[4], wiki_london_data.columns[5] ], axis=1)

In [96]:
df1.head()

London borough       Post_town Postcode district
0              Bexley, Greenwich [7]          LONDON               SE2
1  Ealing, Hammersmith and Fulham[8]          LONDON            W3, W4
2                         Croydon[8]         CROYDON               CR0
3                         Croydon[8]         CROYDON               CR0
4                             Bexley  BEXLEY, SIDCUP         DA5, DA14

let's rename the Postcode district column and the london borough to something simpler

In [97]:
df1.columns = ['borough','town','post_code']
df1

borough            town  post_code
0                Bexley, Greenwich [7]          LONDON        SE2
1    Ealing, Hammersmith and Fulham[8]          LONDON     W3, W4
2                           Croydon[8]         CROYDON        CR0
3                           Croydon[8]         CROYDON        CR0
4                               Bexley  BEXLEY, SIDCUP  DA5, DA14
..                                 ...             ...        ...
526                          Greenwich          LONDON       SE18
527       Sutton, Kingston upon Thames  WORCESTER PARK        KT4
528             Hammersmith and Fulham          LONDON        W12
529                         Hillingdon           HAYES        UB4
530                         Hillingdon    WEST DRAYTON        UB7

[531 rows x 3 columns]

Let's remove the Square brackets [ ] and numbers from the borough column

In [98]:
df1['borough'] = df1['borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1

borough            town  post_code
0                Bexley, Greenwich           LONDON        SE2
1    Ealing, Hammersmith and Fulham          LONDON     W3, W4
2                           Croydon         CROYDON        CR0
3                           Croydon         CROYDON        CR0
4                            Bexley  BEXLEY, SIDCUP  DA5, DA14
..                              ...             ...        ...
526                       Greenwich          LONDON       SE18
527    Sutton, Kingston upon Thames  WORCESTER PARK        KT4
528          Hammersmith and Fulham          LONDON        W12
529                      Hillingdon           HAYES        UB4
530                      Hillingdon    WEST DRAYTON        UB7

[531 rows x 3 columns]

Take the dimension of the dataframe

In [99]:
df1.shape

(531, 3)

We currently have 533 records and 3 columns of our data. It's time to perform Feature Engineering

### Feature Engineering

We can only focusing on the neighbourhoods of London, so performing the changes

In [100]:
df1 = df1[df1['town'].str.contains('LONDON')]
df1

borough                    town post_code
0                Bexley, Greenwich                   LONDON       SE2
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4
6                              City                  LONDON       EC3
7                       Westminster                  LONDON       WC2
9                           Bromley                  LONDON      SE20
..                              ...                     ...       ...
521                       Redbridge                  LONDON  IG8, E18
522       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8
525                          Barnet                  LONDON       N12
526                       Greenwich                  LONDON      SE18
528          Hammersmith and Fulham                  LONDON       W12

[308 rows x 3 columns]

In [101]:
df1.shape

(308, 3)

We now have only 310 rows. We can proceed with our further steps. Getting some descriptive statistics

In [102]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 308 entries, 0 to 528
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   borough    308 non-null    object
 1   town       308 non-null    object
 2   post_code  308 non-null    object
dtypes: object(3)
memory usage: 9.6+ KB


## Geolocations of the London Neighbourhoods

### ArcGis API

We need to get the geographical co-ordinates for the neighbourhoods to plot out map. We will use the arcgis package to do so. 

Arcgis doesn't have a limitation on the number of API calls made so it fits our use case perfectly.

In [103]:
pip install arcgis

You should consider upgrading via the 'c:\users\amana\appdata\local\programs\python\python38\python.exe -m pip install --upgrade pip' command.


In [104]:
from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()

Defining London arcgis geocode function to return latitude and longitude

In [105]:
def get_x_y_uk(address1):
   lat_coords = 0
   lng_coords = 0
   g = geocode(address='{}, London, England, GBR'.format(address1))[0]
   lng_coords = g['location']['x']
   lat_coords = g['location']['y']
   return str(lat_coords) +","+ str(lng_coords)

Checking sample data

In [106]:
c = get_x_y_uk('SE2')

In [107]:
c

'51.492450000000076,0.12127000000003818'

Looks good, We Copy over the postal codes of london to pass it into the geolocator function that we just defined above

In [108]:
geo_coordinates_uk = df1['post_code']    
geo_coordinates_uk

0           SE2
1        W3, W4
6           EC3
7           WC2
9          SE20
         ...   
521    IG8, E18
522         IG8
525         N12
526        SE18
528         W12
Name: post_code, Length: 308, dtype: object

Passing postal codes of london to get the geographical co-ordinates

In [111]:
coordinates_latlng_uk = geo_coordinates_uk.apply(lambda x: get_x_y_uk(x))
coordinates_latlng_uk

0       51.492450000000076,0.12127000000003818
1        51.51324000000005,-0.2674599999999714
6       51.51200000000006,-0.08057999999994081
7       51.51651000000004,-0.11967999999995982
9       51.41009000000008,-0.05682999999993399
                        ...                   
521    51.589770000000044,0.030520000000024083
522      51.50642000000005,-0.1272099999999341
525     51.615920000000074,-0.1767399999999384
526      51.48207000000008,0.07143000000002075
528      51.50645000000003,-0.2369099999999662
Name: post_code, Length: 308, dtype: object

### Latitude

Extracting the latitude from our previously collected coordinates

In [112]:
lat_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[0])
lat_uk

0      51.492450000000076
1       51.51324000000005
6       51.51200000000006
7       51.51651000000004
9       51.41009000000008
              ...        
521    51.589770000000044
522     51.50642000000005
525    51.615920000000074
526     51.48207000000008
528     51.50645000000003
Name: post_code, Length: 308, dtype: object

### Longitude

Extracting the Longitude from our previously collected coordinates

In [113]:
lng_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[1])
lng_uk

0       0.12127000000003818
1       -0.2674599999999714
6      -0.08057999999994081
7      -0.11967999999995982
9      -0.05682999999993399
               ...         
521    0.030520000000024083
522     -0.1272099999999341
525     -0.1767399999999384
526     0.07143000000002075
528     -0.2369099999999662
Name: post_code, Length: 308, dtype: object

We now have the geographical co-ordinates of the London Neighbourhoods.

We proceed with Merging our source data with the geographical co-ordinates to make our dataset ready for the next stage

In [114]:
london_merged = pd.concat([df1,lat_uk.astype(float), lng_uk.astype(float)], axis=1)
london_merged.columns= ['borough','town','post_code','latitude','longitude']
london_merged

borough                    town post_code  \
0                Bexley, Greenwich                   LONDON       SE2   
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4   
6                              City                  LONDON       EC3   
7                       Westminster                  LONDON       WC2   
9                           Bromley                  LONDON      SE20   
..                              ...                     ...       ...   
521                       Redbridge                  LONDON  IG8, E18   
522       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8   
525                          Barnet                  LONDON       N12   
526                       Greenwich                  LONDON      SE18   
528          Hammersmith and Fulham                  LONDON       W12   

     latitude  longitude  
0    51.49245    0.12127  
1    51.51324   -0.26746  
6    51.51200   -0.08058  
7    51.51651   -0.11968  
9    51.41009   -0.05683  
..        ...        ...  
521  51.58977    0.03052  
522  51.50642   -0.12721  
525  51.61592   -0.17674  
526  51.48207    0.07143  
528  51.50645   -0.23691  

[308 rows x 5 columns]

In [115]:
london_merged.dtypes

borough       object
town          object
post_code     object
latitude     float64
longitude    float64
dtype: object

### Co-ordinates for London

Getting the geocode for London to help visualize it on the map

In [116]:
london = geocode(address='London, England, GBR')[0]
london_lng_coords = london['location']['x']
london_lat_coords = london['location']['y']
london_lng_coords

-0.1272099999999341

In [117]:
london_lat_coords

51.50642000000005

## Visualize the Map of London

To help visualize the Map of London and the neighbourhoods in London, we make use of the folium package.

In [120]:
# Creating the map of London
map_London = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)
map_London

# adding markers to map
for latitude, longitude, borough, town in zip(london_merged['latitude'], london_merged['longitude'], london_merged['borough'], london_merged['town']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_London)  
    
map_London

### Venues in London

To proceed with the next part, we need to define Foursquare API credentials.

Using Foursquare API, we are able to get the venue and venue categories around each neighbourhood in London.

In [129]:
CLIENT_ID = 'XOCJVA41PPQHX5ZVPCWPWESBQEJT5J05QUJVEQXPPFROB3YJ'
CLIENT_SECRET = 'KHQOY5O3TIKTEOGRFGQBJ5JUMBLBCLZJMIWGJPCA5NTXABIF'
VERSION = '20180605' # Foursquare API version

Defining a function to get the neraby venues in the neighbourhood. This will help us get venue categories which is important for our analysis

In [130]:
LIMIT=100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            LIMIT
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Getting the venues in London

In [131]:
venues_in_London = getNearbyVenues(london_merged['borough'], london_merged['latitude'], london_merged['longitude'])

Bexley, Greenwich 
Ealing, Hammersmith and Fulham
City
Westminster
Bromley
Islington
Islington
Barnet
Enfield
Wandsworth
Southwark
City
Richmond upon Thames
Barnet
Islington
Wandsworth
Westminster
Bromley
Newham
Ealing
Westminster
Lewisham
Camden
Southwark
Tower Hamlets
Bexley
City
Lewisham
Greenwich
Tower Hamlets
Camden
Haringey
Tower Hamlets
Haringey
Barnet
Brent
Lambeth
Lewisham
Tower Hamlets
Kensington and Chelsea, Hammersmith and Fulham
Brent
Barnet
Barnet
Southwark
Tower Hamlets
Camden
Tower Hamlets
Waltham Forest
Newham
Islington
Richmond upon Thames
Lewisham
Camden
Westminster
Greenwich
Kensington and Chelsea
Barnet
Westminster
Lewisham
Waltham Forest
Hounslow, Ealing, Hammersmith and Fulham
Brent
Barnet
Lambeth, Wandsworth
Islington
Barnet
Merton
Barnet
Westminster
Barnet, Brent, Camden
Lewisham
Bexley
Haringey
Bromley
Tower Hamlets
Newham
Hackney
Islington
Southwark
Lewisham
Brent
Southwark
Ealing
Kensington and Chelsea
Wandsworth
Southwark
Barnet
Newham
Richmond upon Thames


Sampling our data

In [132]:
venues_in_London.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Bexley, Greenwich                 51.49245                  0.12127   
1  Bexley, Greenwich                 51.49245                  0.12127   
2  Bexley, Greenwich                 51.49245                  0.12127   
3  Bexley, Greenwich                 51.49245                  0.12127   
4  Bexley, Greenwich                 51.49245                  0.12127   

                              Venue     Venue Category  
0                      Lesnes Abbey      Historic Site  
1                       Sainsbury's        Supermarket  
2                              Lidl        Supermarket  
3  Abbey Wood Railway Station (ABW)      Train Station  
4                        Costcutter  Convenience Store

In [133]:
venues_in_London.shape

(10351, 5)


Wow, we have scraped together 10351 records for venues. This will definitely make the clustering interesting.



### Grouping by Venue Categories
We need to now see how many Venue Categories are there for further processing

In [134]:
venues_in_London.groupby('Venue Category').max()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                                        
ATM                  Kensington and Chelsea                51.49807   
Accessories Store               Westminster                51.51656   
Adult Boutique                    Islington                51.52969   
African Restaurant              Westminster                51.52587   
American Restaurant          Waltham Forest                51.61780   
...                                     ...                     ...   
Wings Joint          Hammersmith and Fulham                51.54187   
Women's Store                   Westminster                51.55457   
Xinjiang Restaurant               Southwark                51.47480   
Yoga Studio                     Westminster                51.55457   
Zoo Exhibit                          Camden                51.53354   

                     Neighbourhood Longitude                        Venue  
Venue Category                                                             
ATM                                 -0.17404                  NatWest ATM  
Accessories Store                   -0.11968           James Smith & Sons  
Adult Boutique                      -0.08697  Sh! Women's Erotic Emporium  
African Restaurant                  -0.08808           Red Sea Restaurant  
American Restaurant                  0.02795                  Spielburger  
...                                      ...                          ...  
Wings Joint                         -0.19795                     Wingmans  
Women's Store                       -0.11478           Vivien of Holloway  
Xinjiang Restaurant                 -0.09313                    Silk Road  
Yoga Studio                         -0.03558                    yogahaven  
Zoo Exhibit                         -0.14606                Penguin Beach  

[304 rows x 4 columns]

We can see 304 records, just goes to show how diverse and interesting the place is.

### One Hot Encoding 
We need to Encode our venue categories to get a better result for our clustering

In [135]:
London_venue_cat = pd.get_dummies(venues_in_London[['Venue Category']], prefix="", prefix_sep="")
London_venue_cat

ATM  Accessories Store  Adult Boutique  African Restaurant  \
0        0                  0               0                   0   
1        0                  0               0                   0   
2        0                  0               0                   0   
3        0                  0               0                   0   
4        0                  0               0                   0   
...    ...                ...             ...                 ...   
10346    0                  0               0                   0   
10347    0                  0               0                   0   
10348    0                  0               0                   0   
10349    0                  0               0                   0   
10350    0                  0               0                   0   

       American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                        0             0       0                 0   
1                        0             0       0                 0   
2                        0             0       0                 0   
3                        0             0       0                 0   
4                        0             0       0                 0   
...                    ...           ...     ...               ...   
10346                    0             0       0                 0   
10347                    0             0       0                 0   
10348                    0             0       0                 0   
10349                    0             0       0                 0   
10350                    0             0       0                 0   

       Argentinian Restaurant  Art Gallery  ...  Vietnamese Restaurant  \
0                           0            0  ...                      0   
1                           0            0  ...                      0   
2                           0            0  ...                      0   
3                           0            0  ...                      0   
4                           0            0  ...                      0   
...                       ...          ...  ...                    ...   
10346                       0            0  ...                      0   
10347                       0            0  ...                      0   
10348                       0            0  ...                      0   
10349                       0            0  ...                      0   
10350                       0            0  ...                      0   

       Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0                    0           0         0          0            0   
1                    0           0         0          0            0   
2                    0           0         0          0            0   
3                    0           0         0          0            0   
4                    0           0         0          0            0   
...                ...         ...       ...        ...          ...   
10346                0           0         0          0            0   
10347                0           0         0          0            0   
10348                0           0         0          0            0   
10349                0           0         0          0            0   
10350                0           0         0          0            0   

       Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0                  0                    0            0            0  
1                  0                    0            0            0  
2                  0                    0            0            0  
3                  0                    0            0            0  
4                  0                    0            0            0  
...              ...                  ...          ...          ...  
10346              0                    0            0            0  
10347              0   

Adding Neighbourhood into the mix.

In [136]:
London_venue_cat['Neighbourhood'] = venues_in_London['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [London_venue_cat.columns[-1]] + list(London_venue_cat.columns[:-1])
London_venue_cat = London_venue_cat[fixed_columns]

London_venue_cat.head()

Neighbourhood  ATM  Accessories Store  Adult Boutique  \
0  Bexley, Greenwich     0                  0               0   
1  Bexley, Greenwich     0                  0               0   
2  Bexley, Greenwich     0                  0               0   
3  Bexley, Greenwich     0                  0               0   
4  Bexley, Greenwich     0                  0               0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                   0                    0             0       0   
1                   0                    0             0       0   
2                   0                    0             0       0   
3                   0                    0             0       0   
4                   0                    0             0       0   

   Arepa Restaurant  Argentinian Restaurant  ...  Vietnamese Restaurant  \
0                 0                       0  ...                      0   
1                 0                       0  ...                      0   
2                 0                       0  ...                      0   
3                 0                       0  ...                      0   
4                 0                       0  ...                      0   

   Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0                0           0         0          0            0   
1                0           0         0          0            0   
2                0           0         0          0            0   
3                0           0         0          0            0   
4                0           0         0          0            0   

   Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0              0                    0            0            0  
1              0                    0            0            0  
2              0                    0            0            0  
3              0                    0            0            0  
4              0                    0            0            0  

[5 rows x 305 columns]

### Venue categories mean value
We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [137]:
London_grouped = London_venue_cat.groupby('Neighbourhood').mean().reset_index()
London_grouped.head()

Neighbourhood  ATM  Accessories Store  Adult Boutique  \
0                 Barnet  0.0                0.0             0.0   
1  Barnet, Brent, Camden  0.0                0.0             0.0   
2                 Bexley  0.0                0.0             0.0   
3      Bexley, Greenwich  0.0                0.0             0.0   
4     Bexley, Greenwich   0.0                0.0             0.0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                 0.0             0.001761           0.0     0.0   
1                 0.0             0.000000           0.0     0.0   
2                 0.0             0.000000           0.0     0.0   
3                 0.0             0.000000           0.0     0.0   
4                 0.0             0.000000           0.0     0.0   

   Arepa Restaurant  Argentinian Restaurant  ...  Vietnamese Restaurant  \
0               0.0                0.007042  ...               0.001761   
1               0.0                0.000000  ...               0.000000   
2               0.0                0.000000  ...               0.000000   
3               0.0                0.000000  ...               0.000000   
4               0.0                0.000000  ...               0.000000   

   Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0              0.0         0.0       0.0        0.0          0.0   
1              0.0         0.0       0.0        0.0          0.0   
2              0.0         0.0       0.0        0.0          0.0   
3              0.0         0.0       0.0        0.0          0.0   
4              0.0         0.0       0.0        0.0          0.0   

   Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0            0.0                  0.0          0.0          0.0  
1            0.0                  0.0          0.0          0.0  
2            0.0                  0.0          0.0          0.0  
3            0.0                  0.0          0.0          0.0  
4            0.0                  0.0          0.0          0.0  

[5 rows x 305 columns]


Let's make a function to get the top most common venue categories

In [138]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]


There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods.

Creating a function to label the columns of the venue correctly

In [139]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


### Top venue categories

Getting the top venue categories in London

In [140]:
# create a new dataframe for London
neighborhoods_venues_sorted_london = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_london['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighborhoods_venues_sorted_london.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_london.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0                 Barnet           Coffee Shop                  Café   
1  Barnet, Brent, Camden  Gym / Fitness Center        Clothing Store   
2                 Bexley           Supermarket         Historic Site   
3      Bexley, Greenwich               Daycare     Convenience Store   
4     Bexley, Greenwich            Supermarket              Platform   

  3rd Most Common Venue 4th Most Common Venue       5th Most Common Venue  \
0         Grocery Store              Bus Stop                         Pub   
1        Hardware Store           Supermarket                 Zoo Exhibit   
2              Platform     Convenience Store               Train Station   
3           Golf Course          Home Service  Construction & Landscaping   
4     Convenience Store         Train Station               Historic Site   

  6th Most Common Venue 7th Most Common Venue       8th Most Common Venue  \
0    Italian Restaurant           Supermarket                    Pharmacy   
1           Film Studio               Exhibit          Falafel Restaurant   
2              Bus Stop                  Park  Construction & Landscaping   
3         Historic Site              Bus Stop                        Park   
4           Film Studio           Event Space                     Exhibit   

  9th Most Common Venue 10th Most Common Venue  
0    Turkish Restaurant       Sushi Restaurant  
1        Farmers Market   Fast Food Restaurant  
2           Golf Course            Zoo Exhibit  
3           Flea Market            Flower Shop  
4    Falafel Restaurant         Farmers Market